li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java

```diff
@@ -284,6 +300,45 @@ public void flush() {
     _producer.flush();
   }
 
+  /**
+   * This method will flush all the message buffered in producer. The call blocks until timeout.
+   * If the underlying producer doesn't support a bounded flush, it will invoke the {@link #flush()}.
+   */
+  @Override
+  public void flush(long timeout, TimeUnit timeUnit) {
```

Comment:
Should this be synchronized?  If more than one thread calls flush then all of them may need to spin up an extra thread.

Reply:
Probably this needs to see if the producer is already closed.

Reply:
i dont think we should be trying to provide any more concurrency control than the underlying client

Reply:
@smccauliff  If multiple thread call `flush` today, we don't handle it with any additional synchronization than what the underlying producer provides. So, why should that be a concern here? Is it because of the number of threads we end up creating?

Reply:
@smccauliff is the unbounded number of thread creation your only concern here? We can overcome that by limiting the number of threads we spin up and block when the ceiling is reached. This changes the behavior of the flush method as it exists today. Let me know what you think? 
As far as this version is concerned, we don't want to do any incompatible changes.

Reply:
I remember discussing this with you in person several weeks ago and I think we concluded that your implementation was the correct path.

Reply:
Let me take another look tomorrow when I'm more awake.

Reply:
We discussed? I don't remember :( Take a look when you can.